perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_attribute_def_def_o365EmailAddresses_o365mu.java

```diff
@@ -0,0 +1,142 @@
+import static cz.metacentrum.perun.core.impl.Utils.hasDuplicate;
+
+/**
+ * Module for email addresses for Office365 at Masaryk University.
+ * Implements checks for attribute urn:perun:group:attribute-def:def:o365EmailAddresses:o365mu.
+ * <p>
+ * Requirements:
+ * <ul>
+ * <li>type is list</li>
+ * <li>all values are email addresses</li>
+ * <li>must contain at least one value if urn:perun:group:attribute-def:def:adName:o365mu is set</li>
+ * <li>no duplicates among the list values</li>
+ * <li>no duplicates among all values of this attribute and all values of
+ * attribute urn:perun:member:attribute-def:def:o365EmailAddresses:mu</li>
+ * </ul>
+ *
+ * @author Michal Stava &lt;stavamichal@gmail.com&gt;
+ */
+public class urn_perun_group_attribute_def_def_o365EmailAddresses_o365mu extends GroupAttributesModuleAbstract implements GroupAttributesModuleImplApi {
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_group_attribute_def_def_o365EmailAddresses_o365mu.class);
+	static final String ADNAME_ATTRIBUTE = AttributesManager.NS_GROUP_ATTR_DEF + ":adName:o365mu";
+
+	@Override
+	public void checkAttributeSyntax(PerunSessionImpl sess, Group group, Attribute attribute) throws WrongAttributeValueException {
+		log.trace("checkAttributeSyntax(group={},attribute={})", group, attribute);
+		List<String> emails = attribute.valueAsList();
+
+		if (emails == null) return;
+
+		//check syntax of all values
+		for (String email : emails) {
+			Matcher emailMatcher = emailPattern.matcher(email);
+			if (!emailMatcher.matches())
+				throw new WrongAttributeValueException(attribute, group, "Email " + email + " is not in correct form.");
+		}
+
+		//check for duplicities
+		if (hasDuplicate(emails)) {
+			throw new WrongAttributeValueException(attribute, group, "duplicate values");
+		}
+	}
+
+	@Override
+	public void checkAttributeSemantics(PerunSessionImpl sess, Group group, Attribute attribute) throws WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		List<String> emails = attribute.valueAsList();
+
+		if (emails == null) emails = new ArrayList<>();
+		//at least one value if adName is set
+		AttributesManagerBl am = sess.getPerunBl().getAttributesManagerBl();
+		try {
+			if (emails.isEmpty()) {
+				Attribute adName = am.getAttribute(sess, group, ADNAME_ATTRIBUTE);
+				if (adName.getValue() != null) {
+					throw new WrongReferenceAttributeValueException(attribute, adName, group, null, group, null, "at least one email must be defined");
+				}
+
+				return;
+			}
+		} catch (AttributeNotExistsException e) {
+			throw new InternalErrorException(e.getMessage(), e);
+		}
+
+		//check for duplicities among all members and groups
+		AttributesManagerBl attributesManagerBl = sess.getPerunBl().getAttributesManagerBl();
+		Set<Integer> setOfGroupsDefinedByAttribute = attributesManagerBl.getPerunBeanIdsForUniqueAttributeValue(sess, attribute).stream().map(Pair::getLeft).collect(Collectors.toSet());
```

Comment:
If the attribute will be ``unique``, we don't have to check uniqueness through the same attribute, since its checked on inserting the value.

Reply:
Yes, this is not needed here. Only members check against this attribute.